Fix UDP client binding to localhost breaking vote propagation

onnie81 opened the pull request:
Fixes #8339

Commit e1d78e1d04 changed UDP client socket binding from UNSPECIFIED (0.0.0.0) to LOCALHOST (127.0.0.1) to fix test port conflicts in #7490. This broke production validators by preventing vote packets from being sent on actual network interfaces.

**Impact:** Vote latency regressed from ~1 slot to ~2 slots between v2.3.11 and v3.0.4.

**Root cause:** Bisected the entire codebase from v2.3.11 to v3.0.4 and identified commit e1d78e1d04 as the regression point.

**Fix:** Use conditional compilation to bind to LOCALHOST only in tests while restoring UNSPECIFIED binding for production builds.

**Testing:** Tested fix on mainnet validator - vote latency restored to mostly 1s.

onnie81 committed:
Fix UDP client binding to localhost breaking vote propagation (#8339)

t-nelson commented on udp-client/src/lib.rs:
```diff
@@ -63,10 +63,17 @@ pub struct UdpConfig {
     fn new() -> Result<Self, ClientError> {
+        // Use UNSPECIFIED for production validators to bind to all interfaces
+        // Use LOCALHOST only in tests to avoid port conflicts in CI
+        #[cfg(not(test))]
+        let bind_ip = std::net::IpAddr::V4(std::net::Ipv4Addr::UNSPECIFIED);
+        #[cfg(test)]
+        let bind_ip = std::net::IpAddr::V4(std::net::Ipv4Addr::LOCALHOST);
```
i was going to nix this construction as brittle, but this abstraction is so hilariously broken that i think this is actually the least wrong path to backportable fix

lijunwangs replied:
We may simply revert the change of #7490 on this file. i.e. turning back to std::net::IpAddr::V4(std::net::Ipv4Addr::UNSPECIFIED)
so that we can quickly back port 
While addressing the flaky test problem.

onnie81 replied:
Proposing a revert was a bridge too far for me, I focused on trying to keep the logic while solving the perf regression.

if we just revert that file wouldn’t we bringing back potential port conflicts during testing?

lijunwangs replied:
I am not proposing reverting the whole PR. Only the change in this file only.  Having "#[cfg(test)]" is useless and confusing. So instead of having that non-working/misleading config code littered, cleaner to just revert that change. We can fix the test using dev-context-only-utils way in a subsequent PR to unblock this more critical fix.

onnie81 committed:
Use dev-context-only-utils feature instead of cfg(test)

The cfg(test) directive doesn't persist across crate boundaries.
Use dev-context-only-utils feature flag to ensure LOCALHOST binding
in test contexts across all crates.

Tested: Unit tests pass. Release builds verified to compile without
the feature enabled.

lijunwangs commented on udp-client/src/lib.rs:
```diff
@@ -63,10 +63,17 @@ pub struct UdpConfig {
+        // Use LOCALHOST only in dev/test context to avoid port conflicts in CI
+        #[cfg(not(feature = "dev-context-only-utils"))]
+        let bind_ip = std::net::IpAddr::V4(std::net::Ipv4Addr::UNSPECIFIED);
+        #[cfg(feature = "dev-context-only-utils")]
```
This is necessary but not sufficient to fix the flaky port collision issue with unit tests as it must be enabled by the test to use the feature in its dev-dependencies.  But I think we can get this in first as and address the test part in a separate PR as there are quite a few crates need that change.

alexpyattaev replied:
CI should work mostly fine even if it binds to all interfaces. I agree we need to merge this ASAP.

alexpyattaev reviewed (approved):
Thank you very much for your contribution!

alexpyattaev reviewed (approved):
Thank you for your contribution and for spotting this one!

t-nelson commented on udp-client/src/lib.rs:
```diff
@@ -63,10 +63,17 @@ pub struct UdpConfig {
 
 impl NewConnectionConfig for UdpConfig {
     fn new() -> Result<Self, ClientError> {
+        // Use UNSPECIFIED for production validators to bind to all interfaces
+        // Use LOCALHOST only in dev/test context to avoid port conflicts in CI
+        #[cfg(not(feature = "dev-context-only-utils"))]
+        let bind_ip = std::net::IpAddr::V4(std::net::Ipv4Addr::UNSPECIFIED);
+        #[cfg(feature = "dev-context-only-utils")]
+        let bind_ip = std::net::IpAddr::V4(std::net::Ipv4Addr::LOCALHOST);
```
> As an FYI for anyone that hits this later, if you build with `cargo build --release`, then agave-validator will have DCOU _enabled_, and the bind_ip will incorrectly be `Ipv4Addr::LOCALHOST`.

lmao that is its own bug that we need to fix
